### DISCUSSION TITLE
draft framework layout

### DISCUSSION THREAD

**Discussion by thogar-computer** [NERC-EDS/FRAME-FM#17] [Ideas]
This is a starter for 10 on what frame's framework _might_ look like 
```mermaid
classDiagram    
    Dataloader <| -- croissantLoader
    Dataloader <| -- torchGeoLoader
    croissantLoader <| -- GeoCroissantLoader
    Model <|-- NeutalNet
    Model <|-- randomForest
    config <| -- JasmainConfig
    config <| -- IssibardConfig
    crossiantml <| -- GeoCroissantml
    train <| -- trainlocal
    train <| -- traincluster

    

    namespace frame.dataloaders {
        class Dataloader
        class croissantLoader{
            string url
        }
        class torchGeoLoader
        class GeoCroissantLoader
    }
    namespace frame.models {
    class Model {
       +log()
      +summary()
      }
    class NeutalNet{
      int layers
      
    }
    class randomForest {

    }
    }
    namespace frame.configs {    
    class config
    class JasmainConfig
    class IssibardConfig
    }   
    
    namespace frame.untils{
        class crossiantml{
            +validate() : boolen
            +rai_dataPreprocessingProtocol(): void
            +rai_dataBiases(): void
            +rai_dataLimitations(): void
            +rai_dataUseCases(): void
        }
        class GeoCroissantml

    }
    namespace frame.training{
        class train {
            int cpu
            int gpu
            int tpu
        }
        class trainlocal
        class traincluster
    }
    note for config "This is based on hydra"
    note for Model "This is based on pythorch-lighting"
```

**Comment by thogar-computer** [comment]
Looking at this in more detail, there are definitely some defaults when it comes to these types of systems.

- There is usually a main file called train.py (this is the classical main.py and contains the `if __name__ ==` statement for running)
- We are not building a framework for a known outcome (unlike Animoe), so we need to cover the experimentation side of training
- We need a way to see the outputs across different runs
- We want to leave it open to auto-tuning (direction of travel)
 
 Assumption I've made while seeing what this looks like as code
- We want to make sure the framework works locally on Mac/Linux and Windows
- We want to target SLURM

**Reply by thogar-computer** [reply]
Outcome of meeting on 11/11/2025 - SLURM is required for scaling, Most of the work will be via notebooks, running in linux - the `real` training do be done via CLI to SLURM

**Comment by thogar-computer** [comment]
MLflow tracking enabled. View results at: http://127.0.0.1:5000
Generating random image data in: /tmp/tmp13cxd_fb
transforming images...
100%|██████████████████████████████████████████████████████████████████████████████████████████████| 60/60 [00:00<00:00, 380.99it/s]
TinyImageNetModel initialized.
💡 Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
PyTorch Lightning Trainer initialized. Starting training...

| Name      | Type               | Params | Mode |
|-----------|----------------|----------|--------|
| 0 | model     | ResNet             | 5.0 M  | train |
| 1 | train_acc | MulticlassAccuracy | 0      | train |
| 2 | val_acc   | MulticlassAccuracy | 0      | train |

5.0 M     Trainable params
0         Non-trainable params
5.0 M     Total params
20.034    Total estimated model params size (MB)
46        Modules in train mode
0         Modules in eval mode
Sanity Checking: |                                                                                            | 0/? [00:00<?, ?it/s]/mnt/c/work/gemini/frame_project/.linux-venv/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:433: 
/mnt/c/work/gemini/frame_project/.linux-venv/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py:310: The number of training batches (1) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
Epoch 4: 100%|████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  4.18it/s, v_num=0ec0]`Trainer.fit` stopped: `max_epochs=5` reached.                                                                                      
Epoch 4: 100%|████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  0.58it/s, v_num=0ec0]
🏃 View run lightning_run at: http://127.0.0.1:5000/#/experiments/273222403983607525/runs/49e64b4d5f1a4d4b998e90d3858a0ec0
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/273222403983607525
Training complete.
Model checkpoint saved to: /tmp/tmp13cxd_fb/final_model.ckpt
Model logged as an artifact to MLflow.

**Reply by thogar-computer** [reply]
<img width="1966" height="1339" alt="image" src="https://github.com/user-attachments/assets/cd952d37-f1f8-469f-86e6-ce76b069905f" />

**Comment by thogar-computer** [comment]
now see #58
